Add elements

sneakyteak opened the pull request:
Adds a data model for elements, a command for both the full element list and for individual elements, and populates elements for Aphrodite so that this is testable.

sneakyteak committed:
Add elements module and formatted output

Adds a new elements module in data that contains all of the known
elements. Adds new logic to the ability formatter (which is responsible
for the actual bot output for boon commands) to include element info
when it is available. Still works where an element is not defined,
whether because the data is not yet updated or because the boon has no
element (as in the case of infusions). Updates element data in
Aphrodites boons so that this is actually testable end-to-end.

sneakyteak committed:
Adds element and elementList commands

To ship with testable end-to-end functionality, adds an elements entry
to the Aphrodite data. The element command builds a map of each element
to the gods that have boons with that element.

Ellisande commented on src/commands/element.js:
```diff
@@ -0,0 +1,51 @@
+const { Command } = require("./command");
+const { EARTH, WATER, AIR, FIRE, COSMIC, } = require("../data/elements");
```
You actually put the `elements` file in `/data/gods/elements` so this was failing on startup

sneakyteak replied:
🤦‍♂️ thanks!

Ellisande commented on src/commands/element.js:
```diff
@@ -0,0 +1,51 @@
+  example: "water",
+  handler: ({ bot, channelId, commandMatches, logger }) => {
+    const elementName = commandMatches[1];
+    const currentElement = allElements.find((element) => {
+      RegExp(element, "i").test(elementName);
+    });
```
Subtle javascript bug here. If you have an arrow function **without** `{ }` after it, then it implicitly returns, if it has braces, then you need an explicit return. Since you don't return the result of `RegExp(element, "i").test(elementName)` the function always returns undefined, and thus never finds an element.

Without the braces it would've worked as you intended. 

```ts
    const currentElement = allElements.find((element) =>
      element.match(RegExp(elementName, "i"))
    );
```

sneakyteak replied:
ooooh ok, i misunderstood what i read about implicit returns. thanks!

Ellisande commented on src/commands/elementList.js:
```diff
@@ -0,0 +1,19 @@
+const { Command } = require("./command");
+const { EARTH, WATER, AIR, FIRE, COSMIC, } = require("../data/elements");
```
Same bug as above, the file is actually `../data/gods/elements` so this was blowing up

Ellisande committed:
Add tests for the element commands

Ellisande reviewed (changes requested):
Man this looks awesome! Great work considering you were slogging through this all on your own! A couple of minor bugs, one with a incorrect path, and another a subtle javascript bug, that would actually work just fine if it was written in Ruby instead (god love languages). I left the bugs in the PR, because what I added was some example tests that you can use (and add on to!) to verify your stuff is working.

The new tests can be found under `features/elements.feature`. You can run them with the following command:
```shell
npm test features/elements.feature
```

The other feature files are full of other examples, and although they don't work anymore (mostly because their content hasn't been updated) the patterns are still valid.

sneakyteak committed:
Fix path and js bugs in new element code

Incorrect paths to elements data and some bad JS code made this update
DOA. Fixed.

sneakyteak commented:
Thanks for the review and the added tests. Got the test framework up-and-running locally so now I can actually check this code with more tools than my eyes 😅 Looks good to go.
